Content.Shared/Temperature/Systems/SharedTemperatureSystem.cs

```diff
@@ -1,80 +1,431 @@
-                args.CurrentTemperature > threshold && args.LastTemperature < threshold)
-            {
-                ent.Comp.NextSlowdownUpdate = _timing.CurTime + SlowdownApplicationDelay;
-                ent.Comp.CurrentSpeedModifier = modifier;
-                Dirty(ent);
+            // don't do anything if they equalised
+            var diff = Math.Abs(temp.CurrentTemperature - comp.Temperature);
+            if (diff < 0.1f)
+                continue;
+
+            // heat flow in W/m^2 as per fourier's law in 1D.
+            var q = comp.Conductivity * diff / comp.Thickness;
+
+            // convert to J then K
+            var joules = q * comp.Area * frameTime;
+            var degrees = joules / GetHeatCapacity((uid, temp, null));
+            if (temp.CurrentTemperature < comp.Temperature)
+                degrees *= -1;
+
+            // exchange heat between inside and surface
+            comp.Temperature += degrees;
+            Dirty(uid,comp);
+            ForceChangeTemperature((uid, temp), temp.CurrentTemperature - degrees);
+        }
+
+        UpdateDamage(frameTime);
+        UpdateMoveSpeed(frameTime);
+    }
+
+    private void UpdateDamage(float frameTime)
+    {
+        _accumulatedFrametime += frameTime;
+
+        if (_accumulatedFrametime < UpdateInterval)
+            return;
+
+        _accumulatedFrametime -= UpdateInterval;
+
+        if (ShouldUpdateDamage.Count == 0)
+            return;
+
+        foreach (var comp in ShouldUpdateDamage)
+        {
+            MetaDataComponent? metaData = null;
+
+            var uid = comp.Owner;
+            if (Deleted(uid, metaData) || Paused(uid, metaData))
+                continue;
+
+            ChangeDamage(uid, comp);
+        }
+
+        ShouldUpdateDamage.Clear();
+    }
+
+    public void ForceChangeTemperature(Entity<TemperatureComponent?> temperature, float temp)
+    {
+        if (!Resolve(temperature, ref temperature.Comp))
+            return;
+
+        var lastTemp = temperature.Comp.CurrentTemperature;
+        var delta = temperature.Comp.CurrentTemperature - temp;
+        if (delta == 0) //we don't want to dirty or raise events if there was no actual temp change!
+            return;
+        temperature.Comp.CurrentTemperature = temp;
+        Dirty(temperature);
+
+        var ev = new OnTemperatureChangeEvent(temperature.Comp.CurrentTemperature, lastTemp, delta);
+        RaiseLocalEvent(temperature, ev, true);
+    }
+
+    public void ChangeHeat(Entity<TemperatureComponent?> target, float heatAmount, bool ignoreHeatResistance = false)
+    {
+        if (!Resolve(target, ref target.Comp, false))
+            return;
+
+        if (!ignoreHeatResistance)
+        {
+            var modifyChangedTempEvent = new ModifyChangedTemperatureEvent(heatAmount);
+            RaiseLocalEvent(target, modifyChangedTempEvent);
+            heatAmount = modifyChangedTempEvent.TemperatureDelta;
```

Comment:
```suggestion
            heatAmount = modifyChangedTempEvent.HeatDelta;
```
Since we have breaking changes anyways, please rename this variable inside the event. Heat is fundamentally different from temperature (the former is measured in Joule, and the latter in Kelvin) and having the correct physical name is important to make sure it is used correctly by whoever subscribes to it.